src/physics/sitetype.jl

```diff
@@ -0,0 +1,211 @@
+include overloading functions such as `op`,
+`siteinds`, and `state` which generate custom
+operators, Index arrays, and IndexVals associated
+with Index objects having a certain tag.
+
+To make a SiteType type, you can use the string
+macro notation: `SiteType"MyTag"`
+
+To make an SiteType value or object, you can use
+the notation: `SiteType("MyTag")`
+"""
+struct SiteType{T}
+end
+
+SiteType(s::Tag) = SiteType{s}()
+SiteType(s::AbstractString) = SiteType(Tag(s))
+
+macro SiteType_str(s)
+  SiteType{Tag(s)}
+end
+
+"""
+OpName is a parameterized type which allows
+making strings into Julia types for the purpose
+of representing operator names.
+The main use of OpName is overloading the 
+`ITensors.op!` method which generates operators 
+for indices with certain tags such as "S=1/2".
+
+To make a OpName type, you can use the string
+macro notation: `OpName"MyTag"`. 
+
+To make an OpName value or object, you can use
+the notation: `OpName("myop")`
+"""
+struct OpName{Name}
+end
+
+OpName(s::SmallString) = OpName{s}()
+OpName(s::AbstractString) = OpName(SmallString(s))
+
+macro OpName_str(s)
+  OpName{SmallString(s)}
+end
+
+
+# TODO: this should be deprecated in a later 
+# version, but leaving it as a fallback in case
+# any user code uses the old `op` function 
+# pattern
+function old_call_op(s::Index,
+                     opname::AbstractString;
+                     kwargs...)
+  use_tag = 0
+  nfound = 0
+  for n=1:length(tags(s))
+    SType = SiteType{tags(s)[n]}
+    if hasmethod(op,Tuple{SType,Index,AbstractString})
+      use_tag = n
+      nfound += 1
+    end
+  end
+  if nfound == 0
+    throw(ArgumentError("Overload of \"op!\" or \"op\" functions not found for operator name \"$opname\" and Index tags $(tags(s))"))
+  elseif nfound > 1
+    throw(ArgumentError("Multiple tags from $(tags(s)) overload the function \"op\""))
+  end
+
+  st = SiteType(tags(s)[use_tag])
+  return op(st,s,opname;kwargs...)
+end
+
+function _call_op!(s::Index,
+                   opname::AbstractString;
+                   kwargs...)
+  use_tag = 0
+  nfound = 0
+  for n=1:length(tags(s))
+    SType = SiteType{tags(s)[n]}
+    OpN = OpName{SmallString(opname)}
+    if hasmethod(op!,Tuple{ITensor,SType,OpN,Index})
+      use_tag = n
+      nfound += 1
+    end
+  end
+  if nfound == 0
+    # Try fallback to older interface:
+    return old_call_op(s,opname;kwargs...)
+
+    throw(ArgumentError("Overload of \"op!\" functions not found for operator name \"$opname\" and Index tags $(tags(s))"))
+  elseif nfound > 1
+    throw(ArgumentError("Multiple tags from $(tags(s)) overload the function \"op!\" for operator name \"$opname\""))
+  end
+
+  Op = emptyITensor(s',dag(s))
+  st = SiteType(tags(s)[use_tag])
+  opn = OpName(opname)
+  op!(Op,st,opn,s;kwargs...)
+  return Op
+end
+
+function op(opname::AbstractString,
+            s::Index;
+            kwargs...)::ITensor
+
+  opname = strip(opname)
+
+  if opname == "Id"
```

Comment:
Can't this also be replaced by an overload of `op!`?

Reply:
Yes. I was originally thinking that we would eventually deprecate the `op` back-end / implementation functions entirely in favor of the `op!` ones. So I was just leaving these `op` function in place mostly for backwards compatibility and other reasons. But I like your idea of having this system be overloadable at two levels, the `op` level and the `op!` level with the `op` level calling down to the `op!` level and having a generic implementation if it's not overloaded.

So I'm going to redesign the order in which these functions get called, first going through one or two functions called `op` then drilling down into the `op!` overloads.